sda-commons-server-jackson/src/main/java/org/sda/commons/server/jackson/hal/HalLinkProvider.java

```diff
@@ -0,0 +1,129 @@
+ * "contextBasePath/testPath/ResourceID?query=detailed"}
+ */
+public class HalLinkProvider implements Feature {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HalLinkProvider.class);
+
+  @Context private UriInfo uriInfo;
+
+  /**
+   * Creates a {@linkplain HALLink} based on JAX-RS annotated parameters of an interface. This
+   * method requires a second entrypoint with {@linkplain
+   * HalLinkInvocationStateUtility#methodOn(Class)} to process the annotated JAX-RS interface with
+   * the corresponding {@linkplain PathParam} and {@linkplain QueryParam} and the passed arguments
+   * to the proxied method invocation.
+   *
+   * <p>Example:
+   *
+   * <pre>
+   * &#064;Path("")
+   * interface TestApi {
+   *  &#064;Path("/testPath/{testArg}")
+   *  &#064;GET
+   *  String testMethod(@PathParam("testArg") String testArg, @QueryParam("query") query);
+   * }
+   *
+   * HALLink link = halLinkProvider.linkTo(methodOn(TestApi.class).testMethod("ResourceID", "detailed"));
+   * </pre>
+   *
+   * The example would create a HALLink with the following URL: {@code
+   * "contextBasePath/testPath/ResourceID?query=detailed"}
+   *
+   * @param invocation the invocation placeholder.
+   * @return the generated HALLink based on the method invocation
+   * @throws HalLinkMethodInvocationException - If no method invocation is provided via {@linkplain
+   *     HalLinkProvider#methodOn(Class)}
+   */
+  public HALLink linkTo(Object invocation) {
+    final MethodInvocationState methodInvocationState =
+        HalLinkInvocationStateUtility.loadMethodInvocationState();
+    if (invocation != null || !methodInvocationState.isProcessed()) {
+      throw new HalLinkMethodInvocationException("No proxied method invocation processed.");
+    }
+    final UriBuilder uriBuilder =
+        baseUriBuilder()
+            .path(methodInvocationState.getType())
+            .path(methodInvocationState.getType(), methodInvocationState.getInvokedMethod())
+            // Add Path Params from invocation state
+            .resolveTemplates(methodInvocationState.getPathParams());
+    // Add Query Params from invocation state
+    methodInvocationState.getQueryParams().forEach(uriBuilder::queryParam);
+    final HALLink link = new Builder(uriBuilder.build()).build();
+    HalLinkInvocationStateUtility.unloadMethodInvocationState();
+    return link;
+  }
+
+  /**
+   * Creates and returns a proxy instance based on the passed type parameter with a method
+   * invocation handler, which processes and saves the needed method invocation information in the
+   * current thread. Parameters in the afterwards called method of the proxy will be used to resolve
+   * the URI template of the corresponding method.
+   *
+   * <p>After the method invocation of the proxy instance the outer method {@linkplain
+   * HalLinkProvider#linkTo(Object)} will process the derived information of the invocation to *
+   * create the HALLink.
```

Comment:
I think it's worth to mention, that the called method on the API interface must have a return type. As this feature should be used to produce links to resources this should usually be the case. So it does not matter, that `void` return types are not supported.